fix(httpjson): handle message derived query params

diegomarquezp opened the pull request:
Fixes #1783
Some message derived types such as Duration, FieldMask or Int32Value
would not be correctly handled by String.valueOf(). Instead, the
toJson() method is used to make it compliant with the protobuf language
guide

diegomarquezp committed:
fix(httpjson): handle message derived query params

Fixes #1783
Some message derived types such as Duration, FieldMask or Int32Value
would not be correctly handled by String.valueOf(). Instead, the
toJson() method is used to make it compliant with the protobuf languague
guide

blakeli0 commented on gax-httpjson/src/main/java/com/google/api/gax/httpjson/ProtoRestSerializer.java:
```diff
@@ -159,4 +161,19 @@ public String toBody(String fieldName, RequestT fieldValue) {
+   * @param fieldValue a field value to serialize
+   */
+  public String toQueryParamValue(Object fieldValue) {
+    if (fieldValue instanceof GeneratedMessageV3) {
```
I'm not sure we want to handle this for all GeneratedMessageV3, this basically means we want to handle any complex types.

blakeli0 replied:
@vam-google I saw that Noah came up with [a list of wellknownTypes](https://github.com/googleapis/gapic-showcase/blob/fe414784c18878d704b884348d84c68fd6b87466/util/genrest/resttools/populatefield.go#L27), do you think this is a full list for supported types in query params?

diego92sigma6 replied:
I added the well known types to the logic. The check for GeneratedMessageV3 in `toQueryParamValue` will be intersected with serializable message types (e.g. FieldMask) with an extra condition in this if statement - also added a comment line for clarification)

blakeli0 commented:
FYI, a similar [fix](https://github.com/googleapis/gapic-generator-go/pull/1124) is implemented in go.

diegomarquezp committed:
fix(protoparser): decompose messages in query prms

Some message type objects will be passed as query params. These may have
nested properties that will now be generated as ?&foo.bar=1&foo.baz=2

diegomarquezp committed:
test(serializer): add test for complex msg obj

diego92sigma6 commented on gax-httpjson/src/main/java/com/google/api/gax/httpjson/ProtoRestSerializer.java:
```diff
@@ -159,4 +161,19 @@ public String toBody(String fieldName, RequestT fieldValue) {
+  public String toQueryParamValue(Object fieldValue) {
+    if (fieldValue instanceof GeneratedMessageV3) {
+      return toJson(((GeneratedMessageV3) fieldValue).toBuilder(), false)
+          .replaceAll("^\"", "")
```
I did this to get the actual contents of a string value. For example, a FieldMask would serialize into `"a.b.c,a.b"`, but this replace() will get `a.b.c,a.b` without quotes. I will see if there is a better way to ensure sanitization.

diegomarquezp committed:
fix(format): format ProtoRestSerializer files

blakeli0 commented on gax-httpjson/src/main/java/com/google/api/gax/httpjson/ProtoRestSerializer.java:
```diff
@@ -51,6 +58,29 @@
   private final TypeRegistry registry;
 
+  // well-known types obtained from
+  // https://github.com/googleapis/gapic-showcase/blob/fe414784c18878d704b884348d84c68fd6b87466/util/genrest/resttools/populatefield.go#L27
```
I proposed the list from gapic-showcase but I don't think we should just use it, showcase is only used for testing so the list may get out of date. We should probably go through the protobuf doc and add any type that have special handling to the list. For now, since Go already implemented it, we can probably follow what they did [here](https://github.com/googleapis/gapic-generator-go/blob/54f801741cd6c32e67dde2e240c7b0b7af7bd2ee/internal/gengapic/genrest.go#L56).

diego92sigma6 replied:
For sure, I can update the list of well-known types following the [docs list](https://developers.google.com/protocol-buffers/docs/reference/google.protobuf#value), but I'm curious if there is a better way to programatically obtain a list of well-known types so we can update it as a dependency or so.

We have [this documentation entry](https://developers.google.com/protocol-buffers/docs/reference/google.protobuf#value) as a source of well-known types but we may have to manually check from time to time to keep our hardcoded list up to date.
However, I noted that the official definitions such as [duration.proto](https://github.com/protocolbuffers/protobuf/blob/main/src/google/protobuf/duration.proto) would specify a special dotnet package `Google.Protobuf.WellKnownTypes`, whereas in java it falls into the common `com.google.protobuf` package, leaving us with no reliable programatic way to obtain a list of well-known types that would get updated as a dependency. As a last (and bad) source, there is a unit test in the same folder of the duration definition that contains a list of well-known types, but it is a [unit test](https://github.com/protocolbuffers/protobuf/blob/main/src/google/protobuf/unittest_well_known_types.proto), so I don't think we have too many options

blakeli0 commented on gax-httpjson/src/main/java/com/google/api/gax/httpjson/ProtoRestSerializer.java:
```diff
@@ -118,6 +148,18 @@ public void putPathParam(Map<String, String> fields, String fieldName, Object fi
+    for (Map.Entry<FieldDescriptor, Object> fieldEntry :
+        ((GeneratedMessageV3) fieldValue).getAllFields().entrySet()) {
+      Object value = fieldEntry.getValue();
+      putQueryParam(
```
This recursive logic seems complicated and could run into StackOverFlow issue if not careful, also doing all this during runtime could affect the performance for client libraries as well. So now I really think that we should do this `traversing all leaf level fields` part in the generator instead of gax.

diego92sigma6 replied:
Sounds good

blakeli0 commented on gax-httpjson/src/main/java/com/google/api/gax/httpjson/ProtoRestSerializer.java:
```diff
@@ -159,4 +219,23 @@ public String toBody(String fieldName, RequestT fieldValue) {
+          .replaceAll("^\"", "")
+          .replaceAll("\"$", "");
+    }
+    if (fieldValue instanceof ByteString) {
```
Why we need to handle ByteString separately here?

blakeli0 commented on gax-httpjson/src/main/java/com/google/api/gax/httpjson/ProtoRestSerializer.java:
```diff
@@ -159,4 +219,23 @@ public String toBody(String fieldName, RequestT fieldValue) {
+   */
+  public String toQueryParamValue(Object fieldValue) {
+    // This will match with message types that are serializable (e.g. FieldMask)
+    if (fieldValue instanceof GeneratedMessageV3 && !isNonSerializableMessageValue(fieldValue)) {
```
I would just use `Message` instead of `GeneratedMessagev3` in all places since it's more generic. I also found that  we could pass an Enum to query params, where Enum is not of `Message` type, so we may need to handle enum specifically as well.

burkedavison commented on gax-httpjson/src/main/java/com/google/api/gax/httpjson/ProtoRestSerializer.java:
```diff
@@ -127,16 +169,34 @@ public void putPathParam(Map<String, String> fields, String fieldName, Object fi
+      if (isNonSerializableMessageValue(fieldValue)) {
+        putDecomposedMessageQueryParam(fields, fieldName, fieldValue);
+        return;
+      } else {
```
When working with the pattern:
```
if (x is Collection)
  foreach i in x
    // do A(i)
else
  // do A(x)
```
We can often refactor to:
```java
Iterable<?> fieldValues =
    fieldValue instanceof Iterable
        ? (Iterable<?>) fieldValue
        : Collections.singleton(fieldValue);

foreach i in x
  // do A(i)
```
This eliminates the need to study both implementations of A to determine if they're the same or not, with the added benefit of it being easier to test the single code path rather than both. As the complexity of A increases, this simplification becomes more and more beneficial.

burkedavison commented on gax-httpjson/src/main/java/com/google/api/gax/httpjson/ProtoRestSerializer.java:
```diff
@@ -118,6 +148,18 @@ public void putPathParam(Map<String, String> fields, String fieldName, Object fi
+  private void putDecomposedMessageQueryParam(
+      Map<String, List<String>> fields, String fieldName, Object fieldValue) {
+    for (Map.Entry<FieldDescriptor, Object> fieldEntry :
+        ((GeneratedMessageV3) fieldValue).getAllFields().entrySet()) {
```
This cast is only safe due to an instanceof check that occurs elsewhere in the logic. Will the next maintainer know they must call `isNonSerializableMessageValue()` as a check before this method?

I would highly suggest either performing the check explicitly in this method, or changing the parameter from `Object` to `GeneratedMessageV3`, thus forcing the caller of this method to perform the cast.

In general, the type-check (`instanceof`) and type-assertion (`cast`) should always be colocated - as two actions are tightly coupled and will need to change together if one is modified.

burkedavison commented on gax-httpjson/src/main/java/com/google/api/gax/httpjson/ProtoRestSerializer.java:
```diff
@@ -127,16 +169,34 @@ public void putPathParam(Map<String, String> fields, String fieldName, Object fi
+          paramValueList.add(toQueryParamValue(fieldValueItem));
+        }
+      }
+      if (hasProcessedMessage) {
```
Rather than introducing this variable and a secondary exit point for the function, consider instead just performing a `!paramValueList.isEmpty()` check at the end to decide whether to add to the final `fields` map or not.

diegomarquezp committed:
fix(queryparam): use json approach to process msgs

also fixed best practice issues pointed out in last commit's PR

diegomarquezp committed:
fix(queryparam): use numeric value for root enums

enums passed as root object to putQueryParam will now be serialized as
their numeric value

blakeli0 committed:
chore: Refactoring the fix.

diegomarquezp committed:
test(queryparam): atomized tests

Also added tests for serializing objects that contain Any typed
messages. Note that the type registry must have the tested types
beforehand, so they were added in the test class setup

blakeli0 commented on gax-httpjson/src/test/java/com/google/api/gax/httpjson/ProtoRestSerializerTest.java:
```diff
@@ -181,6 +195,75 @@ public void putQueryParam() {
+    expectedFields.put(
+        "object.options.value.value", Arrays.asList("1.000000001s", "2.000000002s", "a.b.c,d.e.f"));
+    // used by JSON parser to obtain descriptors from this type url
+    expectedFields.put(
```
This is an interesting edge case, I don't think we should include these type info in the query params, we probably don't want to configure an Any field as query params as well. Can you do a little more investigation on this?

blakeli0 commented on gax-httpjson/src/test/java/com/google/api/gax/httpjson/ProtoRestSerializerTest.java:
```diff
@@ -53,7 +60,14 @@ public class ProtoRestSerializerTest {
+    // tests with Any type messages require corresponding descriptors in the type registry
+    requestSerializer =
+        ProtoRestSerializer.create(
+            TypeRegistry.newBuilder()
```
Can you check if these descriptors are added to ProtoRestSerializer on generating client libraries? If not, we should added the well knowntypes to the registry as well.

diegomarquezp committed:
test(queryparam): test objects w/ well-known types

diego92sigma6 commented:
The final approach was to rely on JSON serialization to produce fields to be included in the query paramenters. This allows well-known types such as Duration and Timestamp to be serialized into a string with compliant format, plus allowing complex message types to be serialized as tree/array objects that would be recursively processed and added as query parameters
